rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+//! assert_eq!(nd.shape(), Some(vec![2, 2]));
+//! let rnd: ArrayD<f32> = ArrayD::try_from(&nd).unwrap();
+//! assert!(rnd.all_close(&a, 1e-8f32));
+//! ```
+//!
+//! [`Rust's dynamic ndarray`]:https://docs.rs/ndarray/0.12.1/ndarray/
+//! [`copy_from_buffer`]:struct.NDArray.html#method.copy_from_buffer
+//! [`copy_to_ctx`]:struct.NDArray.html#method.copy_to_ctx
+
+use std::convert::TryFrom;
+use std::mem;
+use std::os::raw::c_int;
+use std::ptr;
+use std::slice;
+
+use num_traits::Num;
+use rust_ndarray::{Array, ArrayD};
+
+use ts;
+
+use Error;
+use Result;
+use TVMByteArray;
+use TVMContext;
+use TVMType;
+
+/// See the [`module-level documentation`](../ndarray/index.html) for more details.
+///
+/// Wrapper around TVM array handle.
+#[derive(Debug)]
+pub struct NDArray {
+    pub(crate) handle: ts::TVMArrayHandle,
+    is_view: bool,
+}
+
+impl NDArray {
+    pub(crate) fn new(handle: ts::TVMArrayHandle, is_view: bool) -> Self {
+        NDArray {
+            handle: handle,
+            is_view: is_view,
+        }
+    }
+
+    /// Returns the underlying array handle.
+    #[inline]
+    pub fn as_handle(&self) -> ts::TVMArrayHandle {
+        self.handle
+    }
+
+    #[inline]
+    pub fn is_view(&self) -> bool {
+        self.is_view
+    }
+
+    /// Returns the shape of the NDArray.
+    pub fn shape(&self) -> Option<Vec<usize>> {
+        let arr = unsafe { *(self.handle) };
+        if arr.shape.is_null() || arr.data.is_null() {
+            return None;
+        };
+        let slc = unsafe { slice::from_raw_parts_mut(arr.shape as *mut usize, arr.ndim as usize) };
+        Some(slc.to_vec())
+    }
+
+    /// Returns the total number of entries of the NDArray.
+    #[inline]
+    pub fn size(&self) -> Option<usize> {
+        self.shape().map(|v| v.into_iter().product())
+    }
+
+    /// Returns the context which the NDArray was defined.
+    pub fn ctx(&self) -> TVMContext {
+        let dlctx = unsafe { (*self.handle).ctx };
+        dlctx.into()
+    }
+
+    #[inline]
+    pub fn context(&self) -> TVMContext {
+        self.ctx()
+    }
+
+    /// Returns the type of the entries of the NDArray.
+    pub fn dtype(&self) -> TVMType {
+        let dldtype = unsafe { (*self.handle).dtype };
+        dldtype.into()
+    }
+
+    /// Returns the number of dimensions of the NDArray.
+    pub fn ndim(&self) -> usize {
+        unsafe { (*self.handle).ndim as usize }
+    }
+
+    pub fn strides(&self) -> Option<Vec<usize>> {
+        unsafe {
+            (*self.handle).strides.as_mut().map(|pv| {
+                let sz = self.ndim();
+                let mut v: Vec<usize> = Vec::with_capacity(sz * mem::size_of::<usize>());
+                v.as_mut_ptr()
+                    .copy_from_nonoverlapping(pv as *mut _ as *const _, sz);
+                v.set_len(sz);
+                v
+            })
+        }
+    }
+
+    #[inline]
+    pub fn is_contiguous(&self) -> bool {
+        self.strides().is_none()
```

Comment:
this is almost correct. it's possible to manually set the strides such that the array is contiguous. see https://github.com/dmlc/tvm/blob/master/rust/src/runtime/array.rs#L160

Reply:
right! but I think from the frontend perspective we cannot change that at the moment.